tests/app_net_csm_observability_test.py

```diff
@@ -0,0 +1,615 @@
+SERVER_METRICS = HISTOGRAM_SERVER_METRICS + COUNTER_SERVER_METRICS
+ALL_METRICS = HISTOGRAM_METRICS + COUNTER_METRICS
+
+ClientDeploymentArgs = k8s_xds_client_runner.ClientDeploymentArgs
+KubernetesClientRunner = k8s_xds_client_runner.KubernetesClientRunner
+ServerDeploymentArgs = k8s_xds_server_runner.ServerDeploymentArgs
+KubernetesServerRunner = k8s_xds_server_runner.KubernetesServerRunner
+BuildQueryFn = Callable[[str, str], str]
+ANY = unittest.mock.ANY
+
+
+@dataclasses.dataclass(eq=False)
+class MetricTimeSeries:
+    """
+    This class represents one TimeSeries object
+    from monitoring_v3.ListTimeSeriesResponse.
+    """
+
+    # the metric name
+    name: str
+    # each time series has a monitored resource
+    resource_type: str
+    # each time series has a set of metric labels
+    metric_labels: dict[str, str]
+    # each time series has a set of monitored resource labels
+    resource_labels: dict[str, str]
+    # each time series has a set of data points
+    points: list[monitoring_v3.types.Point]
+
+    @classmethod
+    def from_response(
+        cls,
+        name: str,
+        response: monitoring_v3.types.TimeSeries,
+    ) -> "MetricTimeSeries":
+        return cls(
+            name=name,
+            resource_type=response.resource.type,
+            metric_labels=dict(sorted(response.metric.labels.items())),
+            resource_labels=dict(sorted(response.resource.labels.items())),
+            points=list(response.points),
+        )
+
+    def pretty_print(self) -> str:
+        metric = dataclasses.asdict(self)
+        # too much noise to print all data points from a time series
+        metric.pop("points")
+        return yaml.dump(metric, sort_keys=False)
+
+
+# This class is purely for debugging purposes. We want to log what we see
+# from the Prometheus endpoint before being sent to Cloud Monitoring.
+# Once we determined the root cause of b/323596669 we can remove this
+# class.
+class PrometheusLogger:
+    def __init__(
+        self, k8s_runner: k8s_base_runner.KubernetesBaseRunner, pod_name: str
+    ):
+        logfile_name = (
+            f"{k8s_runner.k8s_namespace.name}_{pod_name}_prometheus.log"
+        )
+        log_path = k8s_runner.logs_subdir / logfile_name
+        self.log_stream: TextIO = open(
+            log_path, "w", errors="ignore", encoding="utf-8"
+        )
+
+    def write(self, line):
+        self.log_stream.write(line)
+        self.log_stream.write("\n")
+        self.log_stream.flush()
+
+    def close(self):
+        self.log_stream.close()
+
+
+class AppNetCsmObservabilityTest(xds_k8s_testcase.AppNetXdsKubernetesTestCase):
+    metric_client: monitoring_v3.MetricServiceClient
+
+    @staticmethod
+    def is_supported(config: skips.TestConfig) -> bool:
+        if config.client_lang == _Lang.CPP and config.server_lang == _Lang.CPP:
+            # CSM Observability Test is only supported for CPP for now.
+            return config.version_gte("v1.62.x")
```

Comment:
No need to check the server version - I've updated the gamma test too

```suggestion
        if config.client_lang == _Lang.CPP:
            return config.version_gte("v1.62.x")
```

Reply:
Done